rust/Earthfile

```diff
@@ -34,21 +41,59 @@ CARGO:
 REMOVE_SOURCE_FINGERPRINTS:
     COMMAND
-    COPY +get-stoml/stoml /tmp/stoml
-    ARG source_libs=$(find . -name Cargo.toml -exec bash -c '/tmp/stoml {} package.name; printf "\n"' \\;)
-    RUN --mount=type=cache,mode=0777,sharing=shared,target=target \
-      find target -name .fingerprint; \
-      fingerprint_folders=$(find target -name .fingerprint) ; \
-      for fingerprint_folder in $fingerprint_folders; do \
-        cd $fingerprint_folder; \
-        for source_lib in $source_libs; do \
-          find . -maxdepth 1 -regex "\./$source_lib-[^-]+" -exec rm -rf {} \; ;\
-        done \
-      done
+    COPY +get-tomljson/tomljson /tmp/tomljson
+    COPY +get-jq/jq /tmp/jq
+    DO +RUN_WITH_CACHE --command="set -e;
+      source_libs=\$(find . -name Cargo.toml -exec bash -c '/tmp/tomljson {} | jq -r .package.name; printf \"\\n\"' \\;) ;
+      fingerprint_folders=\$(find target -name .fingerprint) ;
+      echo \"deleting fingerprints:\";
+      for fingerprint_folder in \$fingerprint_folders; do
+        cd \$fingerprint_folder;
+        for source_lib in \$source_libs; do
+          find . -maxdepth 1 -regex \"\./\$source_lib-[^-]+\" -exec bash -c 'readlink -f {}; rm -rf {}' \; ;
+        done
+      done"
+
+# get-tomljson gets the portable tomljson binary
+get-tomljson:
+    FROM alpine:3.18.3
+    RUN wget -O tomljson.tar.xz https://github.com/pelletier/go-toml/releases/download/v2.1.0/tomljson_2.1.0_linux_amd64.tar.xz; \
+        tar -xf tomljson.tar.xz; \
+        chmod +x tomljson
+    SAVE ARTIFACT ./tomljson tomljson
 
-# get-stoml gets the portable stoml binary
-get-stoml:
+# get-jq gets the portable jq binary
+get-jq:
     FROM alpine:3.18.3
-    RUN wget -O stoml https://github.com/freshautomations/stoml/releases/download/v0.7.1/stoml_linux_amd64; \
-       chmod +x stoml
-    SAVE ARTIFACT ./stoml stoml
+    RUN apk add jq;
+    ARG jq=$(which jq)
+    SAVE ARTIFACT $jq jq
+
+# INSTALL_CARGO_SWEEP installs cargo-sweep if it doesn't exist already
+INSTALL_CARGO_SWEEP:
+    COMMAND
+    DO +RUN_WITH_CACHE --command="set -e;
+        if [ ! -f \$CARGO_HOME/bin/cargo-sweep ]; then
+          cargo install cargo-sweep;
+        fi;"
+
+# RUN_WITH_CACHE runs the passed command with the CARGO caches mounted.
+# Arguments:
+#   - command (required): Command to run, can be any expression.
+#
+# This implementation is not expected to significantly change. Prefer using the
+# `CARGO` UDC if you can, so you can get future improvements transparently.
+RUN_WITH_CACHE:
+    COMMAND
+    ARG command
+    ARG ORIGINAL_CARGO_HOME=$CARGO_HOME
+    ENV CARGO_INSTALL_ROOT=$CARGO_HOME
+    ENV CARGO_HOME="/earthly/.cargo"
+    RUN --mount=type=cache,mode=0777,id="earthly-cargo-cache",sharing=shared,target=/earthly \
+        --mount=type=cache,mode=0777,target=target \
+    set -e; \
+    mkdir -p $CARGO_HOME; \
+    mkdir -p /earthly/bin ;\
+    printf "Running:\n      $command\n"; \
+    eval $command
+    ENV CARGO_HOME=$ORIGINAL_CARGO_HOME
```

Comment:
WDYT of just doing
```sh
    export CARGO_INSTALL_ROOT=$CARGO_HOME
    export CARGO_HOME="/earthly/.cargo"
```
in the `RUN` instead of jumping through these hoops?

Reply:
Wanted the UDC to reset the ENV entry at the end. Notice that the UDC runs in the context of the caller, so having such a side effect is not a good thing

Reply:
Doing `export` as a shell command inside a run doesn't affect the earthly context though, does it?